src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala

```diff
@@ -132,93 +129,73 @@ class ExpressionParser extends JavaTokenParsers {
       // all vertexes of type Workspace filtered for the given namespace and name
       graphPipeline.V("_clazz", classOf[Workspace].getSimpleName).filter(workspaceFunc)
         // then entities from that workspace that match the root entity we are starting at
-        .out(rootType).filter(entityFunc(rootName)),
+        .out(context.rootType).filter(entityFunc),
 
       // text for what the pipeline looks like at this step
-      s"""new GremlinPipeline(graph).V("_entityType", "Workspace").filter(workspaceFunc).out($rootType,$rootName)"""
+      s"""new GremlinPipeline(graph).V("_entityType", "Workspace").filter(workspaceFunc).out(${context.rootType},${context.rootName})"""
     )
   }
 
-  // get a vertex where the name matches the entity we are looking for
-  private def entityFunc(entityName: String) = new PipeFunction[Vertex, java.lang.Boolean] {
-    override def compute(a: Vertex) = {
-      propString(a, "_name") == entityName
-    }
-  }
-
   // add pipe to an entity referenced by the current entity
-  private def entityNameAttributePipeFunc(entityRefName: String)(pipe: PipeType): PipeResult = {
+  private def entityNameAttributePipeFunc(entityRefName: String)(context: ExpressionContext, graphPipeline: PipeType): PipeResult = {
     PipeResult(
       // an entity name is on the outgoing edge label
-      pipe.out(entityRefName),
-      // TODO: is there a way to automate returning this code-string from the code itself? (used for debugging purposes)
+      graphPipeline.out(entityRefName),
       s".out($entityRefName)"
     )
   }
 
-  // TODO: We only support attributes here, not reference types (edges)
-  private def lastAttributePipeFunc(attributeName: String)(pipe: PipeType): FinalResult = {
-    val lastVertices = pipe.toList
+  //TODO: Needs implementation! Return false if reference type.
+  private def isValueType(attributeName: String): Boolean = { true }
 
-    if (lastVertices.size() == 0) {
-      throw new RuntimeException(s"Could not dereference $attributeName because pipe returned no entities")
+  private def lastAttributePipeFunc(attributeName: String)(context: ExpressionContext, graphPipeline: PipeType): FinalResult = {
+    //The below will very likely change. My suspicion is that we can't tell whether the attribute is a value
+    //or reference type without evaluating the pipe passed in and then seeing which of .getProperty or .out works.
+    //Particularly perverse graphs might have a mixture of both! (We should handle such cases cleanly, for some value of "clean".)
+    if (isValueType(attributeName)) {
+      val lastVertices = graphPipeline.toList
+
+      if (lastVertices.size() == 0) {
```

Comment:
not worth changing it, but probably should prefer isEmpty() as it actually reflects intentions and in some collection implementations is optimized.